R/streamable_table.R

```diff
@@ -202,16 +200,47 @@ streamable_base_csv <- function() {
   ## NOTE: write.csv does not permit setting 
   ## `col.names = FALSE``, so cannot omit_header
   write_csv <- function(x, path, omit_header) {
-      utils::write.table(x,
+    utils::write.table(x,
                        file = path, 
                        sep = ",", 
                        quote = TRUE,
                        qmethod = "double",
                        row.names = FALSE,
                        col.names = !omit_header,
                        append = omit_header
-      )
+    )
   }
   streamable_table(read_csv, write_csv, "csv")
 }
 
+
+#' streamable chunked parquet using `arrow`
+#' 
+#' @return a `streamable_table` object (S3)
+#' @export
+#' @seealso [arrow::read_parquet()], [arrow::write_parquet()]
+streamable_parquet <- function() {
+  
+  ## Avoids a hard dependency on arrow for this courtesy function
+  if (!requireNamespace("arrow", quietly = TRUE)) {
+    stop("arrow package must be installed to use parquet",
+         call. = FALSE)
+  }
+  
+  read_parquet <- getExportedValue("arrow", "read_parquet")
+  write_parquet <- getExportedValue("arrow", "write_parquet")
+  
+  read <- function(file, ...) {
+    read_parquet(file, ...)
+  }
+  
+  write <- function(x, path, omit_header = FALSE, ...) {
+    # TODO: use chunk size to automate breaks 
+    write_parquet(x, 
+                  sink = path, 
+                  ...
```

Comment:
How to add chunk_size here?

Reply:
(first apologies if this is all obvious, just stating it for completeness):

So note that the streamable_table$write never "sees" the whole data table.  `ark` calls it once with a header:
https://github.com/ropensci/arkdb/blob/82871d01c8de0aec6d9f064255ceeaa00df2a48a/R/ark.R#L161,

and then again on each chunk: 

https://github.com/ropensci/arkdb/blob/82871d01c8de0aec6d9f064255ceeaa00df2a48a/R/ark.R#-L167-L169

so the size of the `data` blob is actually the size of the chunk.  Note that in this second call, headers are omitted and the data is appended to the previous file.

As you know, parquet doesn't support appending.  Quoting the docs from `arrow::write_parquet()`

> Due to features of the format, Parquet files cannot be appended to. If you want to use the Parquet format but also want the ability to extend your dataset, you can write to additional Parquet files and then treat the whole directory of files as a Dataset you can query. See vignette("dataset", package = "arrow") for examples of this. 

So I think the trick here is that we need to update the filepath so that each subsequent call to write parquet writes to a different file in a given directory.  Maybe `arrow::write_parquet()` will choose a new filename magically if the path (sink) is a directory and not a filepath?  Otherwise we will need to get a bit clever to keep a counter on how many chunks we've done, and so update the filename accordingly....

Reply:
Yes, this is the part that I'm thinking about. arrow::write_parquet() will choose the file name magically if the number of rows is greater than that of `chunk_files` parameter. However, I think I see your point, in order to do this within the framework already provided by arkdb, we need to create this file naming manually and push the files into a directory per table. I'll think about this some more. I'm going to work on this again on the weekend.